lib/api/read.js

```diff
@@ -0,0 +1,47 @@
+var assert = require('assert');
+
+var BigInt = require('../tools/bigint');
+var request = require('../tools/smb2-forge').request;
+var MAX_READ_LENGTH = require('../structures/constants').MAX_READ_LENGTH;
+
+module.exports = function read(file, buffer, offset, length, position, cb) {
+  assert(position !== null, 'null position is not supported');
+
+  var fileSize = BigInt.fromBuffer(file.EndofFile);
+  var bytesRead = 0;
+  var connection = this;
+
+  position = new BigInt(8, position);
+
+  function onRead(err, chunk) {
+    if (err != null) {
+      return cb(err, buffer, bytesRead);
+    }
+
+    chunk.copy(buffer, offset);
+
+    var n = chunk.length;
+    bytesRead += n;
+    offset += n;
+    position = position.add(n);
+
+    readChunk();
+  }
+  function readChunk() {
+    if (bytesRead >= length || position.ge(fileSize)) {
+      cb(null, buffer, bytesRead);
+    }
+    request(
```

Comment:
Do you really intend to call read one last time after the previous test called the final cb ?

Reply:
Nope, that's a mistake, thank you.